### ISSUE TITLE
Prediction Challenge - Cannot Overfit using MTP on a small subset (4 training examples)

### ISSUE THREAD

**Issue by semirelezovikj** [nutonomy/nuscenes-devkit#561]
I am having difficulties trying to overfit the MTPModel for the prediction challenge. After training for 10,000 epochs, the best results I have gotten so far is listed below. Possible issues:

1. The generated ground_truth in the dataset is in the global frame: `ground_truth = self.helper.get_future_for_agent(instance_token, sample_token, 6, in_agent_frame=False)`, since to my understanding the predictions need to be performed in the global frame. Should I do the training based on `in_agent_frame=True` (local frame) and then when calculating the predictions convert the output of MTP model to global frame?
2. My construction of the Prediction challenge is wrong: 
```
 pred_obj = Prediction(
                instance_token[i], 
                sample_token[i], 
                np.array(out[i, :48].reshape(num_modes, 12, 2)),
                probabilities=np.array([out[i, 48], out[i, 49]])
            )
```

Results in prediction challenge: 

```
{
  "MinFDEK": {
    "RowMean": [
      86.68397575502848,
      86.68397575502848,
      86.68397575502848
    ]
  },
  "MinADEK": {
    "RowMean": [
      81.69712783228461,
      81.69712783228461,
      81.69712783228461
    ]
  },
  "MissRateTopK_2": {
    "RowMean": [
      1.0,
      1.0,
      1.0
    ]
  },
  "OffRoadRate": {
    "RowMean": [
      1.0
    ]
  }
}
``` 

# 1. Dataset

The Dataset class being used by the Dataloader: 

```python
import torch
import torch.optim as optim
import torch.nn as nn
from torch.utils.data import DataLoader, Dataset
from typing import List
from nuscenes.prediction.input_representation.static_layers import StaticLayerRasterizer
from nuscenes.prediction.input_representation.agents import AgentBoxesWithFadedHistory
from nuscenes.prediction.input_representation.interface import InputRepresentation
from nuscenes.prediction.input_representation.combinators import Rasterizer
from nuscenes.prediction.models.backbone import ResNetBackbone
from nuscenes.prediction.models.mtp import MTP
from nuscenes.prediction.models.covernet import CoverNet
from nuscenes.prediction import PredictHelper
import numpy as np

class NuscenesDataset(Dataset):
    val_nan_state_vector_cnt = 0
    train_nan_state_vector_cnt = 0
    
    def __init__(self, tokens: List[str], helper: PredictHelper, split: str = 'val'):
        self.tokens = tokens
        self.static_layer_rasterizer = StaticLayerRasterizer(helper)
        self.agent_rasterizer = AgentBoxesWithFadedHistory(helper, seconds_of_history=1)
        self.mtp_input_representation = InputRepresentation(self.static_layer_rasterizer, self.agent_rasterizer, Rasterizer())

        self.split = split  
        self.helper = helper      
        
    def __len__(self):
        return len(self.tokens)

    def __getitem__(self, index: int):
        
        token = self.tokens[index]
        instance_token, sample_token = token.split("_")

        # Wrong: only passed static image without vehicles!
        # image = self.static_layer_representation.
        # make_representation(instance_token, sample_token)
        
        image = self.mtp_input_representation.make_input_representation(instance_token, sample_token)
        
        # Why the permute?
        # Same permute to the tensor done for Covernet as well
        image = torch.Tensor(image).permute(2, 0, 1)
        
        agent_state_vector = torch.Tensor([self.helper.get_velocity_for_agent(instance_token, sample_token),
                                        self.helper.get_acceleration_for_agent(instance_token, sample_token),
                                        self.helper.get_heading_change_rate_for_agent(instance_token, sample_token)])
        
        # agent_state_vector can have nan values
        # 643 / 9041 for val
        # 2297 / 32186 for train
        agent_state_vector = torch.nan_to_num(agent_state_vector)
            
        ground_truth = self.helper.get_future_for_agent(instance_token, sample_token, 6, in_agent_frame=False)
        ground_truth = np.expand_dims(ground_truth, 0)

        return image, agent_state_vector, ground_truth, instance_token, sample_token
```

# 2. MTP Training

```python
from datetime import datetime
import json
import os
Gpu.freegpu()
import torchvision

from nuscenes.prediction.models.mtp import MTP, MTPLoss
NUM_MODES = 2
backbone = ResNetBackbone('resnet50')
model = MTP(backbone, NUM_MODES)
# 
# Gpu.IDS = [1]
device = torch.device('cuda:' + str(Gpu.IDS[0]) if torch.cuda.is_available() else 'cpu')
print('running on ', device)
model = model.to(device)

if torch.cuda.device_count() > 1:
        print("Let's use multiple GPUs:", len(Gpu.IDS), "out of ", torch.cuda.device_count(), "GPUs!")      
        model = nn.DataParallel(model, device_ids=Gpu.IDS)

tokens = train_non_nan[:4]

dataset = NuscenesDataset(tokens, helper)
dataloader = DataLoader(dataset, batch_size=16*len(Gpu.IDS), num_workers=2 * len(Gpu.IDS))

n_iter = 0

minimum_loss = 0

loss_function = MTPLoss(NUM_MODES, 1, 5)
current_loss = 10000
model_dir = make_model_dir()
learning_rates = [0.1]


for lr in learning_rates:
    # optimizer = optim.SGD(model.parameters(), lr=lr, momentum = 0.9)
    optimizer = optim.Adam(model.parameters(), lr=lr)
    print('--------- LEARNING RATE ', lr, '--------------')
    for epoch in range(1, 10001): 
        # print below - only for large dataset
        # print('-> epoch:', epoch)
        for img, agent_state_vector, ground_truth, _, _ in dataloader:            
    #         imshow(torchvision.utils.make_grid(img)) 
            img = img.to(device)
            agent_state_vector = agent_state_vector.to(device)    
            ground_truth = ground_truth.to(device)
            ground_truth = ground_truth.to(torch.float32)
            optimizer.zero_grad()

            prediction = model(img, agent_state_vector)            
    #         print('prediction.shape', prediction.shape, 'ground_truth.shape', ground_truth.shape)
            loss = loss_function(prediction, ground_truth)
            loss.backward()
            optimizer.step()

            current_loss = loss.cpu().detach().numpy()        
            n_iter += 1   
            
            # print(f"Current loss is {current_loss:.4f}, Iter {n_iter}")
            if np.allclose(current_loss, minimum_loss, atol=1e-4):
                print(f"Achieved near-zero loss after {n_iter} iterations.")
                print('saving model at epoch:', epoch)
                state = {
                    'epoch': epoch,
                    'state_dict': model.module.state_dict(),
                    'optimizer': optimizer.state_dict(),
                }
                model_name = 'overfit' + str(epoch)
                model_path = os.path.join(model_dir, model_name + '.t7')
                torch.save(state, model_path)

        if epoch % 200 == 0:
            print('saving model at epoch:', epoch)
            state = {
                'epoch': epoch,
                'state_dict': model.module.state_dict(),
                'optimizer': optimizer.state_dict(),
            }
            model_name = 'overfit' + str(epoch)
            model_path = os.path.join(model_dir, model_name + '.t7')
            torch.save(state, model_path)

        # # if epoch % 50  == 0:         
        # print(f"Current loss is {current_loss:.4f}, Iter {n_iter}, epoch {epoch}")
        # if np.allclose(current_loss, minimum_loss, atol=1e-4):
        #     print(f"Achieved near-zero loss after {n_iter} iterations.")

        n_iter = 0
        print('=> completed epoch ', epoch, 'current_loss', current_loss, ' <=')
```

# 3. Predictions Calculator

```python
    mtp_preds = []
    for img, agent_state_vector, ground_truth, instance_token, sample_token in dataloader:
        print('Iteration', n_iter, 'img.shape', img.shape, 'agent_state_vector.shape', agent_state_vector.shape)
        img = img.to(device)
        agent_state_vector = agent_state_vector.to(device)    
        prediction = model(img, agent_state_vector)

        out = prediction.detach().clone()
        out = out.cpu()  
        print(out.shape, 'out.shape') 


        for i in range(out.shape[0]):
            pred_obj = Prediction(
                instance_token[i], 
                sample_token[i], 
                np.array(out[i, :48].reshape(num_modes, 12, 2)),
                probabilities=np.array([out[i, 48], out[i, 49]])
            )
            mtp_preds.append(pred_obj.serialize())
        
        n_iter += 1
```

**Comment by emwolff** [comment]
@semirelezovikj Sorry for the delay on this. Did you end up resolving your issue?

**Comment by emwolff** [comment]
@semirelezovikj 

https://www.nuscenes.org/nuscenes?tutorial=prediction

1) Yes. Train in the local frame (to normalize) and then just convert to the global frame before submitting.
https://www.nuscenes.org/prediction?externalData=all&mapData=all&modalities=Any
Please see the "Results format" section.

**Comment by semirelezovikj** [comment]
I did - many more epochs were needed (30k+). Since I have your attention now, a quick question: for the prediction challenge, I am working on predicting trajectories by making use of all the agents in a given sample. I was very surprised to note that the training split for example, does not cover all the agents for a given sample. 
Ex:
```
sample: ed2e7d24f2ad41abb3d37cd9a0ca8e89
There are 22 instances for this sample. 
```
However in the training split, you only have 5 instances for that sample: 
```
sample = 'ed2e7d24f2ad41abb3d37cd9a0ca8e89'
instances = ['c3cd792e55e44edf935bc90b489bc587', '0915c23aa6d345e08735185a61ddbdc9', '437fe13d66554c51b1ea89f940e5bf4b', 'ed634e83a0664c59b8d5e7cc84f8ae3e', '38fdf31a4697470e86d0fda3719e426a']
```
Would  making use of the information of all the available instances for a given sample (agent_state_vector, position, potentially rasterized surrounding) be considered as expanding the training dataset? Is this by design on your end, or a buggy omission?

P.S: I will be happy to report this as a separate "issue", or a forum (in which case I apologize for not doing my due dilligence).

**Comment by emwolff** [comment]
@semirelezovikj Note that your learning rate of 0.1 is significantly larger than the 1e-4 that was used in the CoverNet paper (see Sec 4.3). I would expect that you would struggle to converge with 0.1 and SGD w/ momentum. Perhaps using Adam is helping you enough.

The difference should be that only vehicles that move are included in the challenge.

**Comment by emwolff** [comment]
@semirelezovikj For the challenges you should simply use the splits provided.

**Comment by semirelezovikj** [comment]
@emwolff: The reason I am asking is because I am treating the sample as a graph, with the agent as nodes. Correct me if I am wrong, but based on your information, in my example, my sample (the graph) has 22 nodes (the agents), but I only have the ground truth trajectory for 5 of those nodes?

A more important question: when submitting my model to your eval servers, when you pass your list of `instancetoken_sampletoken` of your `test` dataset, I will first convert those tokens to `sample: list of instance tokens` and then treat each sample as graph. The result will be the predicted trajectories and modes for the agents listed in your `test` dataset. 

Hope this setup is acceptable

**Comment by emwolff** [comment]
@semirelezovikj For the purposes of evaluation on the Prediction Challenge, it is okay to use past + current time step for everything, but you should only use future ground truth that we specifically include in the dataset. Does that help clarify?

The eval server is dumb, so regardless of how you make your predictions, just make sure at the end of the day you convert to the expected format.

**Comment by ke-sun** [comment]
@semirelezovikj 
One thing I noticed that in the model definition for MTP and CoverNet, there is no nonlinear activation units (like ReLU) after the backbone. For example, in MTP definiton, `fc1` and `fc2` are of `torch.nn.Linear`.
https://github.com/nutonomy/nuscenes-devkit/blob/9b209638ef3dee6d0cdc5ac700c493747f5b35fe/python-sdk/nuscenes/prediction/models/mtp.py#L72
Effectively, it's one linear layer, which makes the network less expressive. Maybe you could try add some nonlinear activation units in order to make the network overfits (I haven't try this myself :smile:).
